### PR TITLE
Add trace processing and optimize ClickHouse table structures

### PR DESCRIPTION
**Pull request by iuwqyir**
### TL;DR

Added support for trace data storage and retrieval in the ClickHouse and Memory connectors.

### What changed?

- Updated the `Commiter` to handle trace data insertion
- Modified the `ClickHouseConnector` and `MemoryConnector` to include `InsertTraces` and `GetTraces` methods
- Added trace-related functionality to the `IMainStorage` interface
- Updated SQL scripts for creating ClickHouse tables, including a new traces table
- Adjusted engine types and order by clauses for better performance in ClickHouse tables

### How to test?

1. Run the updated SQL scripts to create or modify the ClickHouse tables
2. Verify that trace data is being inserted correctly by checking the `traces` table
3. Test the `GetTraces` method to ensure proper retrieval of trace data
4. Confirm that the `Commiter` is successfully handling trace data alongside blocks, transactions, and logs

### Why make this change?

This change enhances the system's capability to store and retrieve trace data, which is crucial for more comprehensive blockchain analysis. By including trace information, users can gain deeper insights into contract interactions and internal transactions within the blockchain ecosystem.

### CODE REVIEW

File: internal/storage/clickhouse.go [hunk 1]
```diff
@@ -305,72 +305,6 @@ func (c *ClickHouseConnector) GetBlocks(qf QueryFilter) (blocks []common.Block,
 	return blocks, nil
 }
 
-func (c *ClickHouseConnector) DeleteBlocks(blocks []common.Block) error {
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:308]: Why did you remove these?

[the remark this answers is not in the document]
Comment (iuwqyir) [reply internal/storage/clickhouse.go:308]: Let's add them back when we get there. Leaving them here creates clutter and extra maintenance

File: internal/storage/clickhouse.go [hunk 2]
```diff
@@ -607,3 +541,82 @@ func (c *ClickHouseConnector) DeleteBlockData(data []common.BlockData) error {
...
+		)
+		if err != nil {
+			return err
+		}
+	}
+	return batch.Send()
+}
+
+func (c *ClickHouseConnector) GetTraces(qf QueryFilter) (traces []common.Trace, err error) {
+	query := fmt.Sprintf("SELECT * FROM %s.traces FINAL WHERE block_number IN (%s) AND is_deleted = 0%s",
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:581]: is this needed somewhere?

File: internal/storage/memory.go [hunk 3]
```diff
@@ -247,27 +247,6 @@ func getBlockNumbersToCheck(qf QueryFilter) map[string]uint8 {
 	return blockNumbersToCheck
 }
 
-func (m *MemoryConnector) DeleteBlocks(blocks []common.Block) error {
```
Comment (AmineAfia) [inline internal/storage/memory.go:250]: needed for handling reorgs

File: internal/storage/clickhouse.go (hunk 2 above)
Comment (iuwqyir) [reply internal/storage/clickhouse.go:581]: Does the API not support traces?

File: internal/tools/clickhouse_create_logs_table.sql [hunk 4]
```diff
@@ -20,11 +20,6 @@ CREATE TABLE base.logs (
...
     INDEX topic2_idx topic_2 TYPE bloom_filter GRANULARITY 1,
     INDEX topic3_idx topic_3 TYPE bloom_filter GRANULARITY 1,
-) ENGINE = SharedReplacingMergeTree(
-    '/clickhouse/tables/{uuid}/{shard}',
-    '{replica}',
-    insert_timestamp,
-    is_deleted
-)
-ORDER BY (block_number, transaction_hash, log_index) SETTINGS index_granularity = 8192
+) ENGINE = SharedReplacingMergeTree(insert_timestamp, is_deleted)
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_logs_table.sql:23]: what is the implication of this change?

File: internal/tools/clickhouse_create_transactions_table.sql [hunk 5]
```diff
@@ -19,11 +19,5 @@ CREATE TABLE base.transactions (
...
-    '/clickhouse/tables/{uuid}/{shard}',
-    '{replica}',
-    insert_timestamp,
-    is_deleted
-)
-ORDER BY (chain_id, block_number) SETTINGS index_granularity = 8192
-SETTINGS allow_experimental_replacing_merge_with_cleanup = 1;
\ No newline at end of file
+) ENGINE = ReplacingMergeTree(insert_timestamp, is_deleted)
+ORDER BY (chain_id, block_number, hash) SETTINGS allow_experimental_replacing_merge_with_cleanup = 1;
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_transactions_table.sql:23]: is it possible to order by hash? and why do we need ordering by chainId? each chain has it's own table anyways

**Review by AmineAfia — CHANGES_REQUESTED** [review]
have some questions before moving forward

File: internal/tools/clickhouse_create_logs_table.sql (hunk 4 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_logs_table.sql:23]: Removed default index_granularity and `allow_experimental_replacing_merge_with_cleanup` will allow deleting rows marked as deleted

File: internal/storage/clickhouse.go (hunk 2 above)
Comment (AmineAfia) [reply internal/storage/clickhouse.go:581]: no

File: internal/tools/clickhouse_create_transactions_table.sql (hunk 5 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_transactions_table.sql:23]: we have a index for hash.
We have chain_id everywhere in the table schema, that is why. We don't necessarily need separate tables with this schema

Comment (AmineAfia) [reply internal/tools/clickhouse_create_transactions_table.sql:23]: this means the tables are going to have billions of rows and have to handle inserts from multiples indexer instances in parallel

**Review by AmineAfia — APPROVED** [review]

**Comment by iuwqyir** [comment]
### Merge activity

* **Sep 23, 6:18 PM EDT**: @iuwqyir started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/indexer/38).
